File: 0733-flood-fill/0733-flood-fill.cpp
class Solution
{
public:
    bool isValid(vector<vector<int>> &image, int row, int col)
    {
        if (row < 0 || row >= image.size())
            return false;
        else if (col < 0 || col >= image[0].size())
            return false;
        return true;
    }

    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc, int color)
    {
        DFS(image, sr, sc, image[sr][sc], color);
        return image;
    }

    void DFS(vector<vector<int>> &image, int row, int col, int colorToSearch, int color)
    {
        if (!isValid(image, row, col) || image[row][col] == color || image[row][col] != colorToSearch)
            return;
        image[row][col] = color;
        DFS(image, row + 1, col, colorToSearch, color);
        DFS(image, row - 1, col, colorToSearch, color);
        DFS(image, row, col + 1, colorToSearch, color);
        DFS(image, row, col - 1, colorToSearch, color);
    }
};

File: 0733-flood-fill/0733-flood-fill.java
class Solution {
    boolean[][] visited;

    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        visited = new boolean[image.length][image[0].length]; 
        dfs(image, sr, sc, color);
        return image;
    }

    void dfs(int[][] image, int sr, int sc, int color) {
        if (!isValidIndx(image, sr, sc) || visited[sr][sc]) return;

        visited[sr][sc] = true;

        if (isValidIndx(image, sr + 1, sc) && image[sr][sc] == image[sr + 1][sc])
            dfs(image, sr + 1, sc, color);
        if (isValidIndx(image, sr - 1, sc) && image[sr][sc] == image[sr - 1][sc])
            dfs(image, sr - 1, sc, color);
        if (isValidIndx(image, sr, sc + 1) && image[sr][sc] == image[sr][sc + 1])
            dfs(image, sr, sc + 1, color);
        if (isValidIndx(image, sr, sc - 1) && image[sr][sc] == image[sr][sc - 1])
            dfs(image, sr, sc - 1, color);
        image[sr][sc] = color;
    }

    boolean isValidIndx(int[][] image, int sr, int sc) {
        if (sr >= image.length || sr < 0 || sc >= image[0].length || sc < 0) return false;
        return true;
    }
}
